casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala

```diff
@@ -99,149 +172,236 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
-    "(!4) !C"   -> VOLATILE_EVENT             -> merges(S1 | F1, S0, C0),
-    "(!4) C!"   -> VOLATILE_EVENT             -> merges(S0 | F0, C_, S1),
-    "(4!) CX"   -> VOLATILE_EVENT             -> merges(F_ | S0, C_, Nil),
-    "(C!) (C!)" -> PERSISTENT_COULD_NOT_MATCH -> merges(C_ | S0, C_ | S0, Nil)
+    "!X !X"     -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, S0, Nil),
+    "!X !4"     -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, S1, F1),
+    "!X (!4)"   -> VolatileEvent.desc           -> VolatileEvent.mergeable(S0)(S0, F_)(Nil),
+    "!X !C"     -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, S1, C1),
+    "!X (!C)"   -> PersistentCouldMatch.desc    -> PersistentCouldMatch.conflicting(S0)(S0, C_)(Nil),
+    "!X 4X"     -> IncomingCouldMatch.desc      -> IncomingCouldMatch.conflicting(S0, F_, Nil),
+    "!X 4!"     -> HadItsMatch.desc             -> HadItsMatch.mergeable(S0, F_, S0),
+    "!X (4!)"   -> VolatileEvent.desc           -> coveredBy("!X (!4)"),
+    "!X 4!!"    -> HadItsMatch.desc             -> HadItsMatch.mergeable(S0, F_, R0),
+    "!X (4!!)"  -> PersistentNoMatch.desc       -> PersistentNoMatch.mergeable(S0)(F_, R0)(Nil),
+    "!X !!X"    -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, R0, Nil),
+    "!X !!4"    -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, R1, F1),
+    "!X (!!4)"  -> PersistentNoMatch.desc       -> coveredBy("!X (4!!)"),
+    "!X CX"     -> IncomingCouldMatch.desc      -> IncomingCouldMatch.conflicting(S0, C_, Nil),
+    "!X C!"     -> IncomingCouldMatch.desc      -> IncomingCouldMatch.conflicting(S0, C_, S0),
+    "!X (C!)"   -> PersistentCouldMatch.desc    -> PersistentCouldMatch.conflicting(S0)(C_, S0)(Nil),
+    "!4 !4"     -> CouldMatchSame.descConflicts -> CouldMatchSame.conflicting(S0, S1, F_),
+    "!4 !4"     -> CouldMatchSame.descMerges    -> CouldMatchSame.mergeable(S0, S1)(F0, F1),
+    "!4 (!4)"   -> VolatileEvent.desc           -> VolatileEvent.mergeable(S0)(S1, F_)(F0),
+    "(!4) (!4)" -> VolatileEvent.desc           -> VolatileEvent.mergeable(S0, F_)(S0, F_)(Nil),
+    "!4 !C"     -> CouldMatchSame.descMerges    -> CouldMatchSame.mergeable(S0, S1)(F0, C1),
+    "!4 4X"     -> HadItsMatch.desc             -> HadItsMatch.mergeable(S0, F_, F_),
+    "!4 4!"     -> HadItsMatch.desc             -> HadItsMatch.mergeable(S0, F_)(F0, S1),
+    "!4 (4!)"   -> VolatileEvent.desc           -> VolatileEvent.mergeable(S0)(S1, F_)(F0),
+    "!4 4!!"    -> HadItsMatch.desc             -> HadItsMatch.mergeable(S0, F_)(F0, R1),
+    "!4 !!X"    -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, R1, F0),
+    "!4 !!4"    -> CouldMatchSame.descConflicts -> CouldMatchSame.conflicting(S0, R1, F_),
+    "!4 !!4"    -> CouldMatchSame.descMerges    -> CouldMatchSame.mergeable(S0, R1)(F0, F1),
+    "!4 CX"     -> HadItsMatch.desc             -> HadItsMatch.mergeable(S0, C_, F_),
```

Comment:
I know you consider this overkill, but validating the name (`!4 CX`) matches the mergeability case (HadItsMatch) is what was missing for spotting quite a few wrong tests during the review automatically, and who knows how many currently and in the future.

At the same time, parsing the:
```
case class Event(p: Polarity, c: Cardinality) 
case class Example(leftIn: Event, rightIn: Event, leftBase: Option[Event], rightBase: Option[Event])
```
from the Name-s should really be straightforward.

Just as the assertions to be done based on them:
```scala
//same/different polarity:
ex.leftIn.polarity ==/!= ex.rightIn.polarity

//number of matches
(leftBase, rightBase).exists(_.isEmpty) //at most one match
```

I implore you to reconsider this :)

Reply:
Ok, I'm going to reconsider

Reply:
This will have to wait until after rules for peek are implemented

Reply:
I can live with that. Can you though? :) I mean, that's additional safety net that would help both of us when implementing peek. I'm fine either way though :)

Reply:
I treat this self validation as a means to spare the future maintainers some stress. I've been looking at those examples for so long that I know them by heart already, so yes, I can live with that